ntoskrnl/io/pnpmgr/pnpmgr.c

```diff
@@ -1760,6 +1765,81 @@ IopGetParentIdPrefix(PDEVICE_NODE DeviceNode,
     return Status;
 }
 
+static
+ULONG
+IopValidateID(IN PWCHAR Id,
+              IN SIZE_T MaxIdLen,
+              IN ULONG MaxSeparators,
+              IN BOOLEAN IsMultiSz)
+{
+   PWCHAR PtrSymbol;
+   ULONG_PTR StringEnd;
+   WCHAR Symbol;
+   ULONG SeparatorsCount = 0;
+   PWCHAR PtrPrevSymbol = NULL;
+
+   PAGED_CODE();
+
+   StringEnd = (ULONG_PTR)Id + MaxIdLen * sizeof(WCHAR);
+
+   for (PtrSymbol = Id; (ULONG_PTR)PtrSymbol < StringEnd; PtrSymbol++)
+   {
+      Symbol = *PtrSymbol;
+
+      if (Symbol == UNICODE_NULL)
+      {
+         if (!IsMultiSz || (PtrPrevSymbol && PtrSymbol == PtrPrevSymbol + 1))
+         {
+            if (MaxSeparators == SeparatorsCount ||
+                MaxSeparators == MAX_SEPARATORS_MULTISZ)
+            {
+               return ((ULONG_PTR)PtrSymbol - (ULONG_PTR)Id) / 2 + 1;
+            }
+
+            DPRINT("IopValidateID: SeparatorsCount - %d, MaxSeparators - %d\n",
+                   SeparatorsCount,
+                   MaxSeparators);
+
+            // FIXME logging
+            return 0;
+         }
+
+         StringEnd += MaxIdLen * sizeof(WCHAR);
```

Comment:
`StringEnd += MaxIdLen;`
So if I understand correctly, in the case of MULTI_SZ, each (NULL-terminated sub)string has a length of MaxIdLen?
You should add more comments about the format of these strings.

Reply:
As I wrote at the top, IRP_MN_QUERY_ID is documented here https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/irp-mn-query-id
...
"Each hardware ID or compatible ID that a driver returns in this IRP must be less than MAX_DEVICE_ID_LEN characters long. This constant currently has a value of 200 as defined in sdk\inc\cfgmgr32.h. "

Does it make sense to repeat this in the form of a comment?
Each substring in the HardwareID or CompatibleID must be less than 200 characters. For DeviceID and Instance ID this is less. But checking the length for them should not be in this function, but where it checks "If a bus driver supplies globally unique instance IDs for its child devices".

By the way the question on the constant MAX_DEVICE_ID_LEN - I temporarily added it to the beginning of the module (as well as three others). "#include <cfgmgr32.h>" is not correct here.